QuoridorEngine/src/core/QuoridorGame.cs

```diff
@@ -322,37 +419,114 @@ private bool playerCanReachBaseline(bool isWhite)
         /// <param name="row"></param>
         /// <param name="col"></param>
         /// <returns>List of tuples with coordinates of legal neighbours</returns>
-        /// <exception cref="ArgumentException">When provided current
-        /// square's coordinates are invalid</exception>
-        /// 
-        /// [TODO] 
-        /// Handle corner case: other player in neighbour square
-        private List<(int, int)> getLegalNeighbourSquares(int row, int col)
+        private List<(int, int)> getLegalNeighbourSquares(int row, int col, bool currentPlayerIsWhite)
         {
-            if (!board.IsValidPlayerSquare(row, col))
-                throw new ArgumentException("Current square coordinates out of bounds");
+            Debug.Assert(board.IsValidPlayerSquare(row, col));
 
             List<(int, int)> legalNeighbours = new();
 
-            if (board.IsValidPlayerSquare(row - 1, col) &&
-                board.CheckWallPartHorizontal(row, col))
-                legalNeighbours.Add((row - 1, col));
-
-            if (board.IsValidPlayerSquare(row, col - 1) &&
-                board.CheckWallPartVertical(row, col-1))
-                legalNeighbours.Add((row, col - 1));
-
-            if (board.IsValidPlayerSquare(row, col + 1) &&
-                board.CheckWallPartVertical(row, col))
-                legalNeighbours.Add((row, col + 1));
-
-            if (board.IsValidPlayerSquare(row + 1, col) &&
-                board.CheckWallPartHorizontal(row + 1, col))
-                legalNeighbours.Add((row + 1, col));
+            // Square 'DOWN' is valid and not blocked by wall
+            if (board.IsValidPlayerSquare(row - 1, col) && !board.CheckWallPartHorizontal(row, col))
```

Comment:
General comment: maybe we can simplify this later using functions from the Vector2 class. Manhattan distance and vector subtraction functions would make the purpose of this code much clearer if used. However this is not a priority at the moment. Just note this into the code as a TODO comment